### PR TITLE
Issue 129 enable per equation discretisation

### LINKED ISSUE #129: enable per-equation discretiation
depends on #128 

dict of {variable: discretisation} passed as input to base discretisation class. 

The discretiation class passed in the dict handles things like gradient/divergence discretisation + broadcasting

### PR DESCRIPTION
**Pull request by Scottmar93**
# Description
A per **domain** implementation of the spatial methods. This is instead of per equation because of some issues described below. 

This can be made sufficient for different finite elements by adding a polynomial type property to the variables we are solving for. It won't allow for finite volumes and finite elements both on the macroscale but we don't foresee that being something which is needed. 

Fixes #129 

# Details
I have implemented a per domain spatial method assignment. The main reason to do this instead of per equation is that with the per equation method it is not clear how which spatial method to use to discretise the model.variables dictionary which is a dict of useful variables for plotting etc. 

If we figure out a way to implement the model.variables dictionary differently then changing the per domain setup to per equation is quite a small change which only requires a few keys in the discretisation.py file to be edited along with the corresponding tests. 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pybamm/models/base_model.py [hunk 1]
```diff
@@ -67,8 +68,15 @@ def __init__(self):
...
+        self.default_spatial_methods = {
+            "negative electrode": pybamm.FiniteVolume,
+            "separator": pybamm.FiniteVolume,
+            "positive electrode": pybamm.FiniteVolume,
+        }
         self.mesh = pybamm.Mesh(self.default_geometry, submesh_types, submesh_pts)
-        self.default_discretisation = pybamm.FiniteVolumeDiscretisation(self.mesh)
+        self.default_discretisation = pybamm.Discretisation(
+            self.mesh, self.default_spatial_methods
+        )
```
Comment (tinosulzer) [inline pybamm/models/base_model.py]: It could be left to external classes, such as `StandardModelTest`, to call `Mesh(self.default_geometry, self.default_submesh_types, self.default_submesh_pts)` and `Discretisation(mesh, self.default_spatial_methods)` (given that there is now only one `Mesh` class and only one `Discretisation` class).
Otherwise, in every model, if you want to change the default submesh types then you also have to go through and redefine `self.mesh` and then redefine `self.default_discretisation`. 
Similar comment also applies to the line `self.default_parameter_values.process_geometry(self.default_geometry)`

File: pybamm/spatial_methods/finite_volume.py [hunk 2]
```diff
@@ -155,33 +120,33 @@ def gradient_matrix(self, domain):
...
         # Add Neumann boundary conditions if defined
         if symbol.id in boundary_conditions:
             # for the particles there will be a "negative particle" "left" and "right"
             # and also a "positive particle" left and right.
             lbc = boundary_conditions[symbol.id]["left"]
             rbc = boundary_conditions[symbol.id]["right"]
-            discretised_symbol = self.concatenate(lbc, discretised_symbol, rbc)
+            discretised_symbol = pybamm.NumpyModelConcatenation(
+                lbc, discretised_symbol, rbc
+            )
```
Comment (tinosulzer) [inline pybamm/spatial_methods/finite_volume.py]: This should be a `DomainConcatenation` (sorry, realise it's a mistake brought forward from previous implementation)

File: pybamm/spatial_methods/finite_volume.py [hunk 3]
```diff
@@ -155,33 +120,33 @@ def gradient_matrix(self, domain):
...
         domain = symbol.domain
-        # check for
+        # check for spherical domains
         if ("negative particle" or "positive particle") in domain:
 
             # implement spherical operator
             divergence_matrix = self.divergence_matrix(domain)
 
-            submesh = self.mesh.combine_submeshes(*domain)
+            submesh = self._mesh.combine_submeshes(*domain)
```
Comment (tinosulzer) [inline pybamm/spatial_methods/finite_volume.py]: Would you ever have subdomains within the particles? I don't think you'd ever need to combine submeshes, unlike on the macroscale

File: tests/shared.py [hunk 4]
```diff
@@ -52,25 +58,48 @@ def __init__(self, n):
 
         self.mesh = pybamm.Mesh(self.geometry, self.submesh_types, self.submesh_pts)
 
+        self.spatial_methods = {"negative particle": pybamm.FiniteVolume}
```
Comment (tinosulzer) [inline tests/shared.py]:
```suggestion
        self.spatial_methods = {"negative particle": pybamm.SpatialMethodForTesting}
```

File: pybamm/expression_tree/unary_operators.py [hunk 5]
```diff
@@ -172,19 +172,19 @@ class NumpyBroadcast(Broadcast):
...
 
-    def __init__(self, child, domain, mesh):
+    def __init__(self, child, domain, npts):
         super().__init__(child, domain, name="numpy broadcast")
         # determine broadcasting vector size (size 1 if the domain is empty)
         if domain == []:
             self.broadcasting_vector_size = 1
         else:
-            self.broadcasting_vector_size = sum([mesh[dom].npts for dom in domain])
+            self.broadcasting_vector_size = sum([npts[dom] for dom in domain])
```
Comment (tinosulzer) [inline pybamm/expression_tree/unary_operators.py]: This seems like an unnecessary change that obscures what the class is doing, and requires you to artificially create the dictionary `npts` before calling `NumpyBroadcast`. I think that the `npts` attribute of a submesh should always be defined in such a way that `mesh[dom].npts` is what you want to use for broadcasting

File: pybamm/models/base_model.py [hunk 6]
```diff
@@ -54,6 +54,7 @@ def __init__(self):
         )
 
         self.default_geometry = pybamm.Geometry1DMacro()
+
         self.default_parameter_values.process_geometry(self.default_geometry)
```
Comment (tinosulzer) [inline pybamm/models/base_model.py]: See comment below

File: pybamm/models/base_model.py (hunk 1 above)
Comment (Scottmar93) [reply pybamm/models/base_model.py]: Yeah probably should reformulate all this so that just call `tests.shared.testing_disc` and tests.shared.finite_volume_1D_macro_disc` etc. I'll add a new issue to do this as it's a bit of a pain.

File: tests/shared.py (hunk 4 above)
Comment (Scottmar93) [reply tests/shared.py]: Making this small change will break the spherical tests later. This will be dealt with properly in #153 so leaving for now.

File: pybamm/expression_tree/unary_operators.py (hunk 5 above)
Comment (Scottmar93) [reply pybamm/expression_tree/unary_operators.py]: The choice is to:
1) change `SubMesh` so that we add  .npts to each SubMesh within `SpatialMethod` and then pass the new mesh
or 
2) create a dict of npts in each domain and change broadcast to take in npts dict 

I didn't want to change the `SubMesh` within a spatial method so that the same property names refer to slightly different things for different spatial methods. It seems clearer to me to have a submesh.cells and a submesh.edges property rather than renaming one of them submesh.nodes within the a spatial method. Likewise, having submesh.n_cells and a submesh.n_edges instead of submesh.npts.

Comment (tinosulzer) [reply pybamm/expression_tree/unary_operators.py]: ok so rename `nodes` to `cells` and `npts` to `n_cells` and then use `mesh[dom].n_cells`?

File: pybamm/expression_tree/symbol.py [hunk 7]
```diff
@@ -37,6 +37,7 @@ def __init__(self, name, children=[], domain=[]):
             # this also adds copy.copy(child) to self.children
             copy.copy(child).parent = self
         self.domain = domain
+        self.spatial_method = None
```
Comment (tinosulzer) [inline pybamm/expression_tree/symbol.py]:
```suggestion
```
